src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -50494,6 +50499,226 @@ static void G(object? x)
+        public void LambdaWritesDoNotUpdateContainingMethodState()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        string? x = ""hello"";
+        M2(() =>
+        {
+            x = null;
+        });
+        x.ToString();
+
+        x = null;
+        x.ToString(); // 1
+    }
+
+    static void M2(Action a) { }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(16, 9));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LambdaWritesDoNotUpdateContainingMethodState_02()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        M2(() =>
+        {
+            string? y = ""world"";
+            M2(() =>
+            {
+                y = null;
+            });
+            y.ToString();
+
+            y = null;
+            y.ToString(); // 1
+        });
+    }
+
+    static void M2(Action a) { }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (18,13): warning CS8602: Dereference of a possibly null reference.
+                //             y.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(18, 13));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void AnonymousMethodWritesDoNotUpdateContainingMethodState()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        string? x = ""hello"";
+        M2(delegate()
+        {
+            x = null;
+        });
+        x.ToString();
+
+        x = null;
+        x.ToString(); // 1
+    }
+
+    static void M2(Action a) { }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(16, 9));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LocalFunctionWritesDoNotUpdateContainingMethodState()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        string? x = ""hello"";
+        x.ToString();
+
+        M2(local);
+        x.ToString();
+
+        local();
+        x.ToString();
+
+        void local()
+        {
+            x = null;
+        }
+    }
+
+    static void M2(Action a) { }
+}
```

Comment:
Please add a similar test where the local function is defined before `x` is used in the containing method.